File: tests_data/basic/ocaml/math_utils.ml
(* Mathematical utility module in OCaml *)

let rec factorial (n : int) : int =
  if n < 0 then invalid_arg "factorial: negative argument"
  else if n = 0 then 1
  else n * factorial (n - 1)

let rec gcd (a : int) (b : int) : int =
  if b = 0 then abs a
  else gcd b (a mod b)

let is_prime (n : int) : bool =
  if n <= 1 then false
  else
    let rec check d =
      if d * d > n then true
      else if n mod d = 0 then false
      else check (d + 1)
    in
    check 2

let safe_div (x : float) (y : float) : float option =
  if y = 0.0 then None
  else Some (x /. y)

File: tests_data/basic/haskell/math_utils.hs
module MathUtils where

-- Factorial function with pattern matching
factorial :: Integer -> Integer
factorial 0 = 1
factorial n | n > 0 = n * factorial (n - 1)
            | otherwise = error "factorial of negative number"

-- Safe division returning Maybe
safeDiv :: Double -> Double -> Maybe Double
safeDiv _ 0.0 = Nothing
safeDiv x y   = Just (x / y)

-- Check if a number is prime
isPrime :: Integer -> Bool
isPrime n
  | n <= 1 = False
  | otherwise = null [ x | x <- [2 .. floor (sqrt (fromIntegral n))], n `mod` x == 0 ]

-- Fibonacci sequence generator
fib :: Int -> Integer
fib n = fibs !! n
  where fibs = 0 : 1 : zipWith (+) fibs (tail fibs)
